pybossa/redis_lock.py

```diff
@@ -145,13 +145,31 @@ def acquire_lock(self, resource_id, client_id, limit, pipeline=None):
         self._release_expired_locks(resource_id, timestamp)
         if self._redis.hexists(resource_id, client_id):
             return True
-        num_acquired = self._redis.hlen(resource_id)
-        if num_acquired < limit:
-            cache = pipeline or self._redis
-            cache.hset(resource_id, client_id, expiration)
-            cache.expire(resource_id, timedelta(seconds=self._duration))
+
+        pipeline = self._redis.pipeline()
+        if limit == float('inf'):
+            pipeline.hset(resource_id, client_id, expiration)
+            pipeline.expire(resource_id, timedelta(seconds=self._duration))
+            pipeline.execute()
             return True
-        return False
+
+        # Get a mutex lock for updating redis hash with default TTL 3s
+        lock_name = f"{resource_id}_update_mutex"
+        result = False
+
+        try:
+            with self._redis.lock(lock_name, timeout=3, blocking_timeout=1) as mutex:
+                if not mutex.locked():
+                    return False
```

Comment:
Any delay/lag expected due to additional lock here when a task is requested that is still under redundancy limit?

Reply:
Good question. Under the hood, the Redis distributed lock used here is a SET command (calling a lua script to endure the atomicity) with the lock name as the key and a random id as the value(to ensure only the owner can release the lock). Per the doc [here](https://www.datadoghq.com/pdf/Understanding-the-Top-5-Redis-Performance-Metrics.pdf) (page 10), the latency of Redis is about 200us. So the latency added here should be negligible.

Reply:
Thanks for the explanation!